rust/src/concept/value.rs

```diff
@@ -266,26 +264,31 @@ impl Decimal {
     }
 }
 
-impl Display for Decimal {
-    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
-        Debug::fmt(self, f)
+impl fmt::Display for Decimal {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        fmt::Debug::fmt(self, f)
     }
 }
 
-impl Debug for Decimal {
-    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
-        // count number of tailing 0's that don't have to be represented
-        let mut tail_0s = 0;
-        let mut fractional = self.fractional;
-        while fractional % 10 == 0 {
-            tail_0s += 1;
-            fractional /= 10;
-        }
+impl fmt::Debug for Decimal {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        if self.fractional == 0 {
+            write!(f, "{}.0", self.integer_part())?;
+        } else {
+            // count number of tailing 0's that don't have to be represented
+            let mut tail_0s = 0;
+            let mut fractional = self.fractional;
+            while fractional % 10 == 0 {
+                tail_0s += 1;
+                fractional /= 10;
+            }
 
-        // count number of leading 0's that have to be represented
-        let digits = (fractional as f64).log10().floor() as u64;
-        let leading_0s = Self::FRACTIONAL_PART_DENOMINATOR - digits - tail_0s;
-        write!(f, "{}.{:0width$}{}", self.integer_part(), "", digits, width = leading_0s as usize)
+            // count number of leading 0's that have to be represented
+            let digits = fractional.ilog10() as u64; // `fractional` cannot be a power of 10 here, so ilog10 correctly gives the number of digits
```

Comment:
`ilog10` rounds up, which is equivalent to counting decimal digits for positive non-power-of-10 integers.